### PR TITLE
Fixes bug preventing generation of pdfs with non-transparent pngs

### PR DESCRIPTION
**Pull request by kriztynna**
1. Fixes bug
2. Builds out feature spec that would have caught this bug

Solves https://github.com/prawnpdf/prawn/issues/933 and https://github.com/prawnpdf/prawn/issues/937

Already green on Travis: https://travis-ci.org/kriztynna/prawn

### CODE REVIEW

**Comment by cheba** [comment]
Thank you for your contribution. It's much appreciated.

However, I think it's not an optimal solution. For instance, In case of indexed PNG without transparency this code would still generate a mask for the image. A completely opaque mask. It's a redundancy that doesn't contribute to appearance of a document, makes a document heavier and rendering more complicated.

I would suggest completely skip mask generation when there's no transparency. I believe this can be achieved by changing `alpha_channel?` method.

**Comment by kriztynna** [comment]
Thanks @cheba for your prompt and thoughtful response. I've amended the PR per your feedback, almost exactly in the way you suggested. Please check it out when you have a moment.

File: lib/prawn/images/png.rb [hunk 1]
```diff
@@ -108,7 +108,7 @@ def colors
       #
       def split_alpha_channel!
         if alpha_channel?
-          if color_type == 3
+          if color_type == 3 && !@transparency.empty?
```
Comment (cheba) [inline lib/prawn/images/png.rb]: `@transparency.any?` seem a bit shorter and simpler.

But that's just nitpicking.

**Comment by cheba** [comment]
@kriztynna Thank you. This is pretty much what I had in mind.

**Comment by kriztynna** [comment]
Thanks @cheba. I took a closer look at the output of the amended commit, and actually changed it back to something more like the original pull request. While your concerns about performance and such are justified, I think you'll find the difference in output is well justified. For indexed pngs without a transparency palette, we're still going to need to `generate_alpha_channel`, otherwise the output will be garbled, as the attached example files demonstrate.

[Type 3 with @transparent.any? method.pdf](https://github.com/prawnpdf/prawn/files/90671/Type.3.with.transparent.any.method.pdf)
[Type 3 with current PR.pdf](https://github.com/prawnpdf/prawn/files/90672/Type.3.with.current.PR.pdf)

I will say this about the original PR: it didn't handle type 3 pngs with transparency correctly (made them opaque, as you might have surmised). This version no longer does; I've manually checked the output of the acceptance specs for each kind of file, and it's the best solution I've been able to come up with that handles both Type 3's (with and without transparency) correctly.

**Comment by kriztynna** [comment]
Ok, realize I've gone back and forth on this a couple of times, but it's been a learning process. Think I finally understand what you meant about targeting `alpha_channel?` to fix this bug, and amended the commit accordingly.

The attached pdf shows how the current PR would print indexed pngs. It _now_ matches what prawn would have done before the transparency fix was implemented (the last version of my PR was close, but no cigar). And no unnecessary mask generation to boot.

[Type 3 without generating alpha channel.pdf](https://github.com/prawnpdf/prawn/files/91186/Type.3.without.generating.alpha.channel.pdf)

Hope that does the trick.

**Comment by packetmonkey** [comment]
@cheba I'm not as familiar with the png spec as you (or at all really) so once you are good with the branch go ahead and merge it.

Thanks for the help @kriztynna!

**Comment by packetmonkey** [comment]
@kriztynna Because you had a pull request merged, you get commit access! :tada:

[Here are the details.](https://github.com/prawnpdf/prawn/wiki/Contributor-welcome-notes)

Thanks again for helping prawn!!

**Comment by kriztynna** [comment]
Thanks @packetmonkey, I wasn't expecting that! :dancer: 
And thank you to @cheba for reviewing the PR!

**Comment by kriztynna** [comment]
Resolved #938

**Comment by kriztynna** [comment]
Resolved #937

**Pull request merged**